provider/queue.go

```diff
@@ -102,6 +117,10 @@ func (q *Queue) work() {
 		var k datastore.Key = datastore.Key{}
 		var c cid.Cid = cid.Undef
 
+		defer func() {
+			q.done <- struct{}{}
```

Comment:
The idiomatic way to do this is to close the channel. That way, all future reads will return immediately.